New public preview API for confidential ledger code transparency 2025-01-31-preview

ivarprudnikov opened the pull request:
# Data Plane API Specification Update Pull Request

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.


<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## API Info: The Basics

Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document: 
    * Aligning with the latest draft SCITT IETF spec https://datatracker.ietf.org/doc/draft-ietf-scitt-architecture/
    * Aligning with the latest draft SCITT SCRAPI spec https://github.com/ietf-wg-scitt/draft-ietf-scitt-scrapi/blob/main/draft-ietf-scitt-scrapi.md
    * Aligning with the latest draft COSE receipt spec https://datatracker.ietf.org/doc/draft-ietf-cose-merkle-tree-proofs/
* Previous API Spec Doc:
* Updated paths:

As this is a preview API we will deprecate and not support the prior version and all future service apis will follow this newer spec.

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Schedule a data plane REST API spec review](https://aka.ms/azsdk/onboarding/restapischedule)

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories) 
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

fix https://github.com/Azure/azure-rest-api-specs/issues/32559

ivarprudnikov committed:
CTS new preview API version with updated endpoints

ivarprudnikov committed:
merge typespec files into one to make local development possible
generate new openapi specs

ivarprudnikov committed:
adds CBOR errors

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Microsoft.CodeTransparency](https://apiview.dev/Assemblies/Review/dc124dc803df45ac8d694409b75b352f?revisionId=f2682ab394794bd3a36afc054ea77d91)

ivarprudnikov commented:
Private Preview version merged was merged in https://github.com/Azure/azure-rest-api-specs/pull/27916

ivarprudnikov committed:
adds explicit JSON errors
changes transparency-configuration to respond in either CBOR or JSON
adds CBOR errors to entries responses

achamayou commented on specification/confidentialledger/Microsoft.CodeTransparency/main.tsp:
```diff
@@ -28,10 +27,793 @@ namespace Microsoft.CodeTransparency;
+@get
+op getTransparencyConfigJson is Foundations.Operation<
+  {
+    @doc("Accpect header")
```
```suggestion
    @doc("Accept header")
```

achamayou commented on specification/confidentialledger/Microsoft.CodeTransparency/main.tsp:
```diff
@@ -28,10 +27,793 @@ namespace Microsoft.CodeTransparency;
+
+#suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+@added(Versions.`2025-01-31-preview`)
+@doc("Get the service public keys")
```
```suggestion
@doc("Get the public keys used by the service to sign receipts")
```

achamayou commented on specification/confidentialledger/Microsoft.CodeTransparency/main.tsp:
```diff
@@ -28,10 +27,793 @@ namespace Microsoft.CodeTransparency;
+  JsonServerErrors
+>;
+
+#suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+@removed(Versions.`2025-01-31-preview`)
+@doc("Get service parameters")
+@route("/parameters")
+@get
+op getParameters is Foundations.Operation<{}, ParametersResult, ServiceTraits>;
+
+#suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+@removed(Versions.`2025-01-31-preview`)
+@doc("Get the DID configuration file.")
+@route("/.well-known/did.json")
+@get
+op getDidConfig is Foundations.Operation<{}, DidDocument, ServiceTraits>;
```
Do we have a way to/want to mark those as deprecated, and direct usage towards the transparency-configuration endpoint instead?

achamayou commented on specification/confidentialledger/Microsoft.CodeTransparency/main.tsp:
```diff
@@ -28,10 +27,793 @@ namespace Microsoft.CodeTransparency;
+
+#suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+@removed(Versions.`2025-01-31-preview`)
+@doc("Get Entry with or without receipt embedded in its unprotected header")
```
Do we want to clarify that Entries are always Signed Statements?

ivarprudnikov replied:
changed to entry -> signed statement

achamayou commented on specification/confidentialledger/Microsoft.CodeTransparency/main.tsp:
```diff
@@ -28,10 +27,793 @@ namespace Microsoft.CodeTransparency;
+}
+
+@added(Versions.`2025-01-31-preview`)
+@doc("Response of entry submission if the response can be served immediatelly, see IETF SCITT SCRAPI draft")
```
```suggestion
@doc("Response of entry submission if the response can be served immediately, see IETF SCITT SCRAPI draft")
```

achamayou commented on specification/confidentialledger/Microsoft.CodeTransparency/main.tsp:
```diff
@@ -28,10 +27,793 @@ namespace Microsoft.CodeTransparency;
+
+@doc("A ledger receipt, see IETF SCITT draft")
+model ReceiptEntry {
+  @doc("The MIME content type for raw receipt is application/cbor.")
```
What are raw receipts? Is that the old receipt format?

ivarprudnikov replied:
removed word `raw` but this refers to the older API as it contains `@removed`

ivarprudnikov committed:
typos

ivarprudnikov committed:
text improvements

ivarprudnikov committed:
add examples to the new api version

ivarprudnikov committed:
update examples

ivarprudnikov committed:
json reformat

ivarprudnikov committed:
cspell fixes

ivarprudnikov committed:
add new api to the readme

ivarprudnikov committed:
update examples

ivarprudnikov committed:
refactor sbor error from object to file

ivarprudnikov committed:
error model change and warning suppresions

ivarprudnikov committed:
maxLength constraint added to openapi

ivarprudnikov committed:
example json tweaks

ivarprudnikov committed:
mark older api endpoints as deprecated in the description

ivarprudnikov committed:
remove serviceIssuer and serviceFeed

vimauro commented on specification/confidentialledger/Microsoft.CodeTransparency/main.tsp:
```diff
@@ -28,10 +27,764 @@ namespace Microsoft.CodeTransparency;
+  transactionIds: string[];
+
+  @doc("Path to the next page of results")
+  @nextLink
```
nit: getting this error while compiling it locally: 

` error ambiguous-symbol: "nextLink" is an ambiguous name between TypeSpec.nextLink, Azure.Core.nextLink. Try using fully qualified name instead: TypeSpec.nextLink, Azure.Core.nextLink`

vimauro replied:
Using   _@Azure.Core.nextLink_ seems to work

ivarprudnikov replied:
Had to use `@TypeSpec.nextLink` because  `@Azure.Core.nextLink` does not compile in typespec

ivarprudnikov committed:
adds jwks cspell

ivarprudnikov committed:
add jwks to cspell

ivarprudnikov committed:
rename generated methods to avoid "Changed" in them

ivarprudnikov committed:
change @nextLink to @TypeSpec.nextLink as otherwise .net sdk does not generate

ivarprudnikov committed:
rename example files due to method name change in newer version

ivarprudnikov committed:
added missing method name renames to the examples

ivarprudnikov committed:
more missing example changes and references

ivarprudnikov committed:
Merge branch 'main' into confidentialledger/cts/2025-01-31-preview

vimauro commented on specification/confidentialledger/Microsoft.CodeTransparency/main.tsp:
```diff
@@ -28,10 +27,764 @@ namespace Microsoft.CodeTransparency;
+@doc("Get signed statement with or without receipt embedded in its unprotected header")
+@sharedRoute
+@route("/entries/{entryId}")
+op getEntry is Foundations.Operation<
```
as tested during SDK generation, this should be 

```
op getEntry is Foundations.Operation<
  {
    @doc("Accept header")
    @header
    accept: "application/cose";
  } &
  EmbedReceiptParameter,
  SignedStatement | TransparentStatement,
  ServiceTraits,
  AnyCborError
>;

```

ivarprudnikov replied:
Thanks, updated

ivarprudnikov committed:
use explicit accept header for the sdk geneartor to skip accept parameter

ivarprudnikov committed:
add missing accept parameter in the example file to getentry

ivarprudnikov commented:
/azp run

ivarprudnikov commented:
/azp run TypeSpec APIView

ivarprudnikov committed:
remove deprecated methods as recommended in the api review as this was never GA'd and is in preview

ivarprudnikov committed:
simplify the api to only the minimum of required endpoints

ivarprudnikov committed:
update compiled swagger

ivarprudnikov committed:
adds content types to examples

ivarprudnikov committed:
typo

ivarprudnikov committed:
remove unused examples

ivarprudnikov committed:
lowercase acceptparameter

ivarprudnikov committed:
add api version parameter to make swagger model validator happy

ivarprudnikov committed:
add missing 202 response in OperationStatusPending

ivarprudnikov committed:
prettify comments to make spellchecker happier

ivarprudnikov committed:
model validation fixes for getoperation example

ivarprudnikov committed:
swagger prettier check

ivarprudnikov committed:
prettify getoperation example

ivarprudnikov committed:
use Azure.Core.Versions.v1_0_Preview_1 instead of v2

ivarprudnikov committed:
Merge branch 'main' into confidentialledger/cts/2025-01-31-preview

ivarprudnikov committed:
Merge branch 'main' into confidentialledger/cts/2025-01-31-preview

ivarprudnikov committed:
Merge remote-tracking branch 'origin/main' into confidentialledger/cts/2025-01-31-preview

ivarprudnikov committed:
remove unnecessary supressions for 2025-01-31-preview/cts.json

ivarprudnikov committed:
add LroExtensionback to supressions

mikeharder commented:
@danieljurek: Please investigate the failure of `Swagger ApiDocPreview`.

maririos commented:
Based on conversations internally in the team I `approved` the suppressions added

ivarprudnikov committed:
Merge branch 'main' into confidentialledger/cts/2025-01-31-preview

ivarprudnikov committed:
Merge branch 'main' into confidentialledger/cts/2025-01-31-preview
